### PR TITLE
Initial version of reactive DB2 SQL client

### PR DESCRIPTION
**Pull request by aguibert**
This is an initial rough cut of what I've been working on over the past few months. It is NOT complete by any means, and there is still a lot of work that needs to be done including things like:
- Proper error path handling
- Implement all TCK tests
- Stress testing
- Secure authentication mechanisms

A full list of remaining work can be seen here:
https://github.com/eclipse-vertx/vertx-sql-client/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+%5BDB2%5D

This initial PR does build and test cleanly, doing a `mvn clean install` should download and start a DB2 container (via Testcontainers) and have about 80 tests that run cleanly.

Most of the code is in the `io.vertx.db2client.impl.drda` package, which handles reading and writing the DRDA protocol (the network protocol used to communicate with DB2). This code is ported from the Apache Derby project (Apache V2 licensed) and reworked to be independent of JDBC. The eventual goal here is to separate out the DRDA package into its own separate repository so it can be used to build other SQL clients that interface with DB2. For ease of development initially however, I plan to iterate on it in this repository.

### CODE REVIEW

**Comment by aguibert** [comment]
Also CC @BillyYccc for review

**Comment by aguibert** [comment]
Travis build is red but this is because there is a MySQL/MariaDB test failing `MySQLTLSTest.testConnFail`. The new DB2 tests are all green

**Comment by BillyYccc** [comment]
> Travis build is red but this is because there is a MySQL/MariaDB test failing `MySQLTLSTest.testConnFail`. The new DB2 tests are all green

Yes this is a regression, see #504

File: vertx-db2-client/src/test/java/io/vertx/db2client/tck/DB2TextDataTypeDecodeTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Override
+    protected void initConnector() {
+        connector = ClientConfig.CONNECT.connect(vertx, rule.options());
+    }
+
+    @Test
+    public void testDecimal(TestContext ctx) {
+        // In DB2 NUMERIC type == DECIMAL type so always return BigDecimal
+        testDecodeGeneric(ctx, "test_decimal", BigDecimal.class, BigDecimal.valueOf(12345));
```
Comment (BillyYccc) [inline vertx-db2-client/src/test/java/io/vertx/db2client/tck/DB2TextDataTypeDecodeTest.java]: can we use the `Numeric` Vert.x type to replace `BigDecimal` like other projects do? it's useful to be used in a polyglot environment.

Comment (aguibert) [reply vertx-db2-client/src/test/java/io/vertx/db2client/tck/DB2TextDataTypeDecodeTest.java]: good point, I've pushed another commit that resolves this so now `testDecimal` and `testNumeric` are not overridden in the DB2 TCK tests

**Comment by aguibert** [comment]
Closes #485 
Closes #486 
Closes #487 
Closes #488

File: pom.xml [hunk 2]
```diff
@@ -59,7 +59,7 @@
     <!-- We skip sources jar generation as we do it with the assembly plugin to have greater control over the content -->
     <source.skip>true</source.skip>
     <kotlin.version>1.3.0</kotlin.version>
-    <testcontainers.version>1.12.0</testcontainers.version>
+    <testcontainers.version>1.12.4</testcontainers.version>
```
Comment (vietj) [inline pom.xml]: We cannot change the version without a works-with CQ - can you revert this and make a separate PR for which we will create the works-with CQ separately ?

File: vertx-db2-client/README.adoc [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+```
+mvn test
+```
+
+== Testing with an external database
+
+If you want to use an external database, you must set the following environment variables:
+
+- DB2_HOST: The hostname of the machine where DB2 is
```
Comment (vietj) [inline vertx-db2-client/README.adoc]: are those the commonly used env names ?

for PG we do use `PGHOST`, etc... because those are _officially_ used see https://www.postgresql.org/docs/9.6/static/libpq-envars.html

just asking a question here

File: vertx-db2-client/pom.xml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      <optional>true</optional>
+    </dependency>
+    <dependency>
+      <groupId>io.vertx</groupId>
+      <artifactId>vertx-sql-client</artifactId>
+    </dependency>
+    
+    <!-- Temp dependency for cross-checking w/ JDBC -->
+    <dependency>
+        <groupId>com.ibm.db2</groupId>
```
Comment (vietj) [inline vertx-db2-client/pom.xml]: we need a CQ for this - if that's only used in tests then it's a works-with CQ

File: vertx-db2-client/scripts/db2.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+echo "Starting DB2 database with:"
+echo "    user=db2user"
+echo "    pass=db2pass"
+echo "  dbname=vertx_db"
+echo "    port=50000"
+echo ""
+echo "### DB will be usable when you see the following message ###"
+echo "    (*) Setup has completed."
+echo ""
+docker run --ulimit memlock=-1:-1 -it --rm=true --memory-swappiness=0 \
```
Comment (vietj) [inline vertx-db2-client/scripts/db2.sh]: good idea to use `-it` and `--rm=true` it makes the script easier to use

File: vertx-db2-client/scripts/db2.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+echo "Starting DB2 database with:"
+echo "    user=db2user"
+echo "    pass=db2pass"
+echo "  dbname=vertx_db"
+echo "    port=50000"
+echo ""
+echo "### DB will be usable when you see the following message ###"
```
Comment (vietj) [inline vertx-db2-client/scripts/db2.sh]: you should add a line of text to say that the DB will run until the user hits ctrl-c to shutdow the docker container

File: vertx-db2-client/src/main/java/examples/DB2ClientExamples.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+import io.vertx.sqlclient.Tuple;
+
+/**
+ * To Run DB2 in a docker container thaty is compatible with this sample, run
+ * the script at scripts/db2.sh
+ */
+public class DB2ClientExamples {
+	
+	static final String HOST = "localhost";
+//	static final String HOST = "192.168.1.22";
```
Comment (vietj) [inline vertx-db2-client/src/main/java/examples/DB2ClientExamples.java]: commented hardcoded code

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/CloseConnectionCommandCodec.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    try {
+        ByteBuf packet = allocateBuffer();
+        DRDAQueryRequest closeCursor = new DRDAQueryRequest(packet);
+        // TODO: @AGG track open cursors somehow so they can be closed here
+        //closeCursor.buildCLSQRY(SectionManager.INSTANCE.getDynamicSection(), encoder.socketConnection.database(), 1); // @AGG guessing 1 on queryInstanceId
+        closeCursor.buildRDBCMM();
+        closeCursor.completeCommand();
+        sendNonSplitPacket(packet);
+    } catch (Exception e) {
+        e.printStackTrace();
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/CloseConnectionCommandCodec.java]: log

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/CloseConnectionCommandCodec.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  void decodePayload(ByteBuf payload, int payloadLength) {
+	  if (LOG.isDebugEnabled())
+		  LOG.debug("disconnect reply");
+      try {
+          DRDAQueryResponse closeCursor = new DRDAQueryResponse(payload);
+          //closeCursor.readCursorClose();
+          closeCursor.readLocalCommit();
+      } catch (Exception e) {
+          payload.clear();
+          e.printStackTrace();
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/CloseConnectionCommandCodec.java]: log

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/DB2Codec.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  
+  public static String dumpBuffer(ByteBuf buffer) {
+      return dumpBuffer(buffer, buffer.readableBytes());
+  }
+  
+  public static String dumpBuffer(ByteBuf buffer, int length) {
+      return dumpBuffer(buffer, length, -1);
+  }
+  
+  public static String dumpBuffer(ByteBuf buffer, int length, int calloutIndex) {
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/DB2Codec.java]: Netty has an utility for this you can reuse `ByteBufUtils#appendPrettyHexDump` that is used by Netty's `LoggingHandler`

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/DB2Decoder.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            ctx.decodePayload(payload, payloadLength);
+        } catch (Throwable t) {
+            int i = payload.readerIndex();
+            payload.readerIndex(startIndex);
+            LOG.error("FATAL: Error parsing buffer at index " + i + " / 0x" + Integer.toHexString(i) + "\n" + 
+                    DB2Codec.dumpBuffer(payload, payloadLength, i), t);
+            ctx.completionHandler.handle(CommandResponse.failure(t));
+        }
+        payload.clear();
+        payload.release();
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/DB2Decoder.java]: releasing should be in a finally block

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/ExtendedQueryCommandCodec.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        ByteBuf packet = allocateBuffer();
+        DRDAQueryRequest openQuery = new DRDAQueryRequest(packet);
+        String dbName = encoder.socketConnection.database();
+        int fetchSize = 0; // TODO @AGG get fetch size from config
+        Tuple params = cmd.params();
+        Object[] inputs = new Object[params.size()];
+        for (int i = 0; i < params.size(); i++)
+            inputs[i] = params.getValue(i);
+        
+        if (DRDAQueryRequest.isQuery(cmd.sql())) {
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/ExtendedQueryCommandCodec.java]: so the protocol used depends on the interpretation of the SQL query ?

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/InitialHandshakeCommandCodec.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+                    0, // targetCmnappc, (not used)
+                    0, // targetXamgr,
+                    0, // targetSyncptmgr,
+                    0, // targetRsyncmgr,
+                    CCSIDManager.TARGET_UNICODE_MGR // targetUnicodemgr
+            );
+            cmd.buildACCSEC(TARGET_SECURITY_MEASURE, this.cmd.database(), null);
+            cmd.completeCommand();
+        } catch (SQLException e) {
+            e.printStackTrace();
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/InitialHandshakeCommandCodec.java]: log

**Comment by vietj** [comment]
code needs to follow https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/CCSIDManager.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.vertx.db2client.impl.drda;
+
+import java.nio.charset.Charset;
+
+public class CCSIDManager {
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/CCSIDManager.java]: it seems this class should rather be called CCSIDConstants

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/CCSIDManager.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    
+    public static final int CCSID_EBCDIC = 500; // 0x01F4
+    public static final int CCSID_UTF8 = 1208; // 0x04B8
+    public static final int TARGET_UNICODE_MGR = CCSID_UTF8;
+    
+    public static final Charset EBCDIC = Charset.forName("CP1047");
+    public static final Charset UTF8 = Charset.forName("UTF-8");
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/CCSIDManager.java]: use `StandardCharsets.UTF_8`

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/ColumnMetaData.java [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.vertx.db2client.impl.drda;
+
+public class ColumnMetaData {
+    
+    public int columns_;
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/ColumnMetaData.java]: why naming with trailing underscores ?

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/Cursor.java [hunk 17]
```diff
@@ -0,0 +1,2479 @@
...
+    private boolean qryclsimpEnabled_;
+
+    //-----------------------------constants--------------------------------------
+
+    //---------------------constructors/finalizer---------------------------------
+
+    Cursor() {
+    }
+
+    Cursor(int qryprctyp)  //protocolType, CodePoint.FIXROWPRC | CodePoint.LMTBLKPRC
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/Cursor.java]: doc explaining what `qryprctyp` means

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/DRDAQueryRequest.java [hunk 18]
```diff
@@ -0,0 +1,1862 @@
...
+    private final HashMap<Integer, Object> promototedParameters_ = new HashMap<>();
+    
+    public DRDAQueryRequest(ByteBuf buffer) {
+        super(buffer);
+    }
+    
+    /**
+     * @return True if the SQL is a query (i.e. SELECT) and false otherwise
+     */
+    public static boolean isQuery(String sql) {
```
Comment (vietj) [inline vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/DRDAQueryRequest.java]: should the string be checked for whitespaces ?

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/ExtendedQueryCommandCodec.java (hunk 12 above)
Comment (aguibert) [reply vertx-db2-client/src/main/java/io/vertx/db2client/impl/codec/ExtendedQueryCommandCodec.java]: The protocol used is always DRDA, but yes we need to send different commands over the wire for an update vs. query statement. I believe that we technically could use the same commands for both, but if we flow the more specific commands then the DB can make some optimizations on its end

File: vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/ColumnMetaData.java (hunk 16 above)
Comment (aguibert) [reply vertx-db2-client/src/main/java/io/vertx/db2client/impl/drda/ColumnMetaData.java]: this is how the code was from Apache Derby, I'm trying to modify the code as little as possible for starters while we try to get everything working so that it's easier to diff the original code. After we get everything working well I plan to come back to the drda package and do some major refactoring

File: pom.xml (hunk 2 above)
Comment (aguibert) [reply pom.xml]: Submitted the CQ for testcontainers update here: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=21539

File: vertx-db2-client/pom.xml (hunk 4 above)
Comment (aguibert) [reply vertx-db2-client/pom.xml]: Submitted a works-with CQ here: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=21540
